xarray/core/parallel.py

```diff
@@ -0,0 +1,333 @@
+        #                   [name2, variable2],
+        #                   ...]
+        # which is passed to dict and then to Dataset
+        data_vars = []
+        coords = []
+
+        for name, variable in dataset.variables.items():
+            # make a task that creates tuple of (dims, chunk)
+            if dask.is_dask_collection(variable.data):
+                # recursively index into dask_keys nested list to get chunk
+                chunk = variable.__dask_keys__()
+                for dim in variable.dims:
+                    chunk = chunk[chunk_index_dict[dim]]
+
+                chunk_variable_task = ("tuple-" + dask.base.tokenize(chunk),) + v
+                graph[chunk_variable_task] = (
+                    tuple,
+                    [variable.dims, chunk, variable.attrs],
+                )
+            else:
+                # non-dask array with possibly chunked dimensions
+                # index into variable appropriately
+                subsetter = {}
+                for dim in variable.dims:
+                    if dim in chunk_index_dict:
+                        which_chunk = chunk_index_dict[dim]
+                        subsetter[dim] = slice(
+                            chunk_index_bounds[dim][which_chunk],
+                            chunk_index_bounds[dim][which_chunk + 1],
+                        )
+
+                subset = variable.isel(subsetter)
+                chunk_variable_task = (name + dask.base.tokenize(subset),) + v
+                graph[chunk_variable_task] = (
+                    tuple,
+                    [subset.dims, subset, subset.attrs],
+                )
+
+            # this task creates dict mapping variable name to above tuple
+            if name in dataset._coord_names:
+                coords.append([name, chunk_variable_task])
+            else:
+                data_vars.append([name, chunk_variable_task])
+
+        from_wrapper = (gname,) + v
+        graph[from_wrapper] = (
+            _wrapper,
+            func,
+            (Dataset, (dict, data_vars), (dict, coords), dataset.attrs),
+            input_is_array,
+            args,
+            kwargs,
+        )
+
+        # mapping from variable name to dask graph key
+        var_key_map = {}  # type: Dict[Hashable, str]
+        for name, variable in template.variables.items():
+            if name in indexes:
+                continue
+            gname_l = "%s-%s" % (gname, name)
+            var_key_map[name] = gname_l
+
+            key = (gname_l,)  # type: Tuple[Any, ...]
+            for dim in variable.dims:
+                if dim in chunk_index_dict:
+                    key += (chunk_index_dict[dim],)
+                else:
+                    # unchunked dimensions in the input have one chunk in the result
+                    key += (0,)
+
+            graph[key] = (operator.getitem, from_wrapper, name)
+
+    graph = HighLevelGraph.from_collections(gname, graph, dependencies=[dataset])
```

Comment:
Just verifying here.  Are there likely to be any dask objects in any of `args` or `kwargs`?  People might do something like put a `ds.mean()` value in there somewhere or a dask.delayed value.  Maybe we save this for a future PR, but it might be good to sift through args and kwargs and either verify that they aren't dask collections or else include them in the computation.

Reply:
good catch! Raising an error now.